File: src/main/java/no/nav/syfo/controller/PDFRestController.java
package no.nav.syfo.controller;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.domain.Soknad;
import no.nav.syfo.domain.dto.PDFTemplate;
import org.springframework.http.*;
import org.springframework.stereotype.Controller;
import org.springframework.web.client.RestTemplate;

import static org.springframework.http.HttpStatus.OK;

@Slf4j
@Controller
public class PDFRestController {
    private RestTemplate restTemplate;

    public PDFRestController(RestTemplate restTemplate) {
        this.restTemplate = restTemplate;
    }

    public byte[] getPDF(Soknad soknad, PDFTemplate template) {
        final String url = "http://pdf-gen.default/api/v1/genpdf/syfosoknader/" + template;

        final HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        final HttpEntity<Soknad> entity = new HttpEntity<>(soknad, headers);

        final ResponseEntity<byte[]> result = restTemplate.exchange(url, HttpMethod.POST, entity, byte[].class);

        if (result.getStatusCode() != OK) {
            throw new RuntimeException("getPDF feiler med HTTP-" + result.getStatusCode() + " for søknad om utenlandsopphold med id: " + soknad.getSoknadsId());
        }

        return result.getBody();
    }
}

File: src/main/kotlin/no/nav/syfo/controller/PDFRestController.kt
package no.nav.syfo.controller

import no.nav.syfo.domain.Soknad
import no.nav.syfo.domain.dto.PDFTemplate
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus.OK
import org.springframework.http.MediaType
import org.springframework.retry.annotation.Backoff
import org.springframework.retry.annotation.Retryable
import org.springframework.stereotype.Controller
import org.springframework.web.client.RestTemplate

@Controller
class PDFRestController(
    private val restTemplate: RestTemplate,
    @Value("\${pdfgen.url}") private val pdfgenUrl: String
) {

    @Retryable(backoff = Backoff(delay = 5000))
    fun getPDF(soknad: Soknad, template: PDFTemplate): ByteArray? {
        val url = "$pdfgenUrl/api/v1/genpdf/syfosoknader/" + template.endpoint

        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_JSON

        val entity = HttpEntity(soknad, headers)

        val result = restTemplate.exchange(url, HttpMethod.POST, entity, ByteArray::class.java)

        if (result.statusCode != OK) {
            throw RuntimeException("getPDF feiler med HTTP-" + result.statusCode + " for søknad om utenlandsopphold med id: " + soknad.soknadsId)
        }

        return result.body
    }
}
